cranelift/assembler-x64/src/api.rs

```diff
@@ -0,0 +1,155 @@
+
+/// Describe how an instruction is emitted into a code buffer.
+pub trait CodeSink {
+    /// Add 1 byte to the code section.
+    fn put1(&mut self, _: u8);
+
+    /// Add 2 bytes to the code section.
+    fn put2(&mut self, _: u16);
+
+    /// Add 4 bytes to the code section.
+    fn put4(&mut self, _: u32);
+
+    /// Add 8 bytes to the code section.
+    fn put8(&mut self, _: u64);
+
+    /// Inform the code buffer of a possible trap at the current location;
+    /// required for assembling memory accesses.
+    fn add_trap(&mut self, code: TrapCode);
+
+    /// Return the byte offset of the current location in the code buffer;
+    /// required for assembling RIP-relative memory accesses.
+    fn current_offset(&self) -> u32;
+
+    /// Inform the code buffer of a use of `label` at `offset`; required for
+    /// assembling RIP-relative memory accesses.
+    fn use_label_at_offset(&mut self, offset: u32, label: Label);
+
+    /// Return the label for a constant `id`; required for assembling
+    /// RIP-relative memory accesses of constants.
+    fn get_label_for_constant(&mut self, id: Constant) -> Label;
+}
+
+/// Provide a convenient implementation for testing.
+impl CodeSink for Vec<u8> {
+    fn put1(&mut self, v: u8) {
+        self.extend_from_slice(&[v]);
+    }
+
+    fn put2(&mut self, v: u16) {
+        self.extend_from_slice(&v.to_le_bytes());
+    }
+
+    fn put4(&mut self, v: u32) {
+        self.extend_from_slice(&v.to_le_bytes());
+    }
+
+    fn put8(&mut self, v: u64) {
+        self.extend_from_slice(&v.to_le_bytes());
+    }
+
+    fn add_trap(&mut self, _: TrapCode) {}
+
+    fn current_offset(&self) -> u32 {
+        self.len().try_into().unwrap()
+    }
+
+    fn use_label_at_offset(&mut self, _: u32, _: Label) {}
+
+    fn get_label_for_constant(&mut self, c: Constant) -> Label {
+        Label(c.0)
+    }
+}
+
+/// Wrap [`CodeSink`]-specific labels.
+#[derive(Debug, Clone, Arbitrary)]
+pub struct Label(pub u32);
+
+/// Wrap [`CodeSink`]-specific constant keys.
+#[derive(Debug, Clone, Arbitrary)]
+pub struct Constant(pub u32);
+
+/// Wrap [`CodeSink`]-specific trap codes.
+#[derive(Debug, Clone, Copy, Arbitrary)]
+pub struct TrapCode(pub NonZeroU8);
+
+/// A table mapping `KnownOffset` identifiers to their `i32` offset values.
+///
+/// When encoding instructions, Cranelift may not know all of the information
+/// needed to construct an immediate. Specifically, addressing modes that
+/// require knowing the size of the tail arguments or outgoing arguments (see
+/// `SyntheticAmode::finalize`) will not know these sizes until emission.
+///
+/// This table allows up to do a "late" look up of these values by their
+/// `KnownOffset`.
+pub trait KnownOffsetTable: Index<KnownOffset, Output = i32> {}
+impl KnownOffsetTable for Vec<i32> {}
+impl KnownOffsetTable for [i32; 2] {}
+
+/// A `KnownOffset` is a unique identifier for a specific offset known only at
+/// emission time.
+pub type KnownOffset = usize;
```

Comment:
Can we make this a newtype wrapper (`pub struct KnownOffset(usize)`) instead?

Reply:
Leaving it as a `type` enables us to just tag on the `Index<KnownOffset, ...>` to the trait and no implementation is needed for vectors and slices. Since I wasn't really in love with this part (see my initial comments above) I took a pretty minimal approach. Let's figure out the big picture for this and then we can decide whether to flesh this out more.